src/services/mapCode.ts

```diff
@@ -0,0 +1,322 @@
+    isTypeElement,
+    isWhileStatement,
+    LanguageServiceHost,
+    Mutable,
+    Node,
+    NodeArray,
+    or,
+    ScriptTarget,
+    some,
+    SourceFile,
+    Statement,
+    SyntaxKind,
+    textChanges,
+    TextSpan,
+    TypeElement,
+    UserPreferences,
+} from "./_namespaces/ts.js";
+import { ChangeTracker } from "./textChanges.js";
+
+/** @internal */
+export function mapCode(
+    sourceFile: SourceFile,
+    contents: string[],
+    focusLocations: TextSpan[][] | undefined,
+    host: LanguageServiceHost,
+    formatContext: formatting.FormatContext,
+    preferences: UserPreferences,
+): FileTextChanges[] {
+    return textChanges.ChangeTracker.with(
+        { host, formatContext, preferences },
+        changeTracker => {
+            const parsed = contents.map(parse);
+            const flattenedLocations = focusLocations && flatten(focusLocations);
+            for (const nodes of parsed) {
+                placeNodeGroup(
+                    sourceFile,
+                    changeTracker,
+                    nodes,
+                    flattenedLocations,
+                );
+            }
+        },
+    );
+}
+
+/**
+ * Tries to parse something into either "top-level" statements, or into blocks
+ * of class-context definitions.
+ */
+function parse(content: string): NodeArray<Node> {
+    // We're going to speculatively parse different kinds of contexts to see
+    // which one makes the most sense, and grab the NodeArray from there. Do
+    // this as lazily as possible.
+    const nodeKinds = [
+        {
+            parse: () =>
+                createSourceFile(
+                    "__mapcode_content_nodes.ts",
+                    content,
+                    ScriptTarget.Latest,
+                    /*setParentNodes*/ true,
+                ),
+            body: (sf: SourceFile) => sf.statements,
+        },
+        {
+            parse: () =>
+                createSourceFile(
+                    "__mapcode_class_content_nodes.ts",
+                    `class __class {\n${content}\n}`,
+                    ScriptTarget.Latest,
+                    /*setParentNodes*/ true,
+                ),
+            body: (cw: SourceFile) => (cw.statements[0] as ClassLikeDeclaration).members,
+        },
+    ];
+
+    const parsedNodes = [];
+    for (const { parse, body } of nodeKinds) {
+        const sourceFile = parse();
+        const bod = body(sourceFile);
+        if (bod.length && sourceFile.parseDiagnostics.length === 0) {
+            // If we run into a case with no parse errors, this is likely the right kind.
+            return bod;
+        }
+        // We only want to keep the ones that have some kind of body.
+        else if (bod.length) {
+            // Otherwise, we'll need to look at others.
+            parsedNodes.push({ sourceFile, body: bod });
+        }
+    }
+    // Heuristic: fewer errors = more likely to be the right kind.
+    const { body } = parsedNodes.sort(

```

Comment:
We are using parseDiagnostics here for heuristics so i strongly feel we should be using ScriptTarget and LanguageVersion of sourceFile for the temp sourcefiles that are created.